### DISCUSSION TITLE
Git LFS locking on non-latest version

### DISCUSSION THREAD

**Discussion by SalmaFarghaly** [git-lfs/git-lfs#4990] [General] [answered]
Hello,

How does git lfs behave if I run "git lfs lock" on non-latest version of a file ? . I can see from the experiment with command line that it doesn't refuse to give me lock. But conflicts are reported during the push. I want to avoid any merges and keep the editing serialized also I may not remember to run "git pull" before running "git lfs lock".  Is there a way to make git lfs tell me that I am locking a non-latest version of the file when I run "git lfs lock" ?

**Comment by chrisd8088** [comment]
Hey, I'm sorry you're having trouble.  Would you be able to provide an example session which demonstrates the problem you're having?  For instance, showing the commands you run, then the commands someone else runs which updates the remote despite your lock, then the command you run which show the conflicts?

The reason I ask is that so far as I know locks are only taken against paths, not specific revisions or commits (although they can include an optional ref name).  Here's the lock API request [structure](https://github.com/git-lfs/git-lfs/blob/4bd61bfc495042a759da6e7f13e5a839fd36eec1/locking/api.go#L30-L36) and here's what it looks like in practice:
```
{"path":"123.bin","ref":{"name":"refs/heads/main"}}
```

(That's despite what the original locking API proposal [says](https://github.com/git-lfs/git-lfs/blob/4bd61bfc495042a759da6e7f13e5a839fd36eec1/docs/proposals/locking_api.md#request), which doesn't reflect the simplifications made in #1909.)

I can, for instance, lock a file which doesn't even exist on the remote:
```
$ touch new.txt
$ git lfs lock new.txt
Locked new.txt
$ git lfs unlock -f new.txt
Warning: unlocking with uncommitted changes because --force
Unlocked new.txt
```

So to lock a given path one shouldn't need the latest commit, and if that were so, one would end up with a race -- someone could always push a new revision just after you'd fetched the latest but before you'd taken the lock.

That all said, if you have a reproducible simple example, that would help to understand the problem you're encountering.

**Comment by SalmaFarghaly** [comment]
Hello @chrisd8088 , Thank you for responding.
Here is a scenario to demonstrate the case I am talking about. Let's assume there is a file called "file.txt" that I want to avoid the merge conflicts to it. 

And I need to modify "file.txt" But that file has some changes on remote that I don't have in my local working copy (i.e my local head is behind the remote master and I don't have any extra changes that aren't on the remote ) and the "file.txt" isn't locked by someone else. Now I have forget to run "git pull" command to get latest changes of "file.txt" and I have run "git lfs lock file.txt". Now I will take the lock and start editing, After I finish editing I have run "git add file.txt" and "git commit --m "commit"" . Now when I run "git push origin master" the push will be rejected because I have forgot to run "git pull" at first. merge conflicts will arise now when I try to run "git pull" command.

So, I am asking if there is some way that git lfs can alert me before locking files that I haven't their latest changes. That is to avoid the possible merge conflicts in the future.

**Comment by chrisd8088** [comment]
OK, thanks for the details.  The Git LFS locking API is separate from Git (and typically implemented on a different server by forges like GitHub), so I don't think there's a simple way for it to automatically achieve what you're hoping.  For one thing, the Git server usually doesn't check with the Git LFS server to see if there's a lock or not; the locking is only advisory and only implemented by Git LFS clients.

You might be better off creating a small local Git extension or a shell script alias to do what you want.  For instance, here's a little script which I put in a file I named `/usr/local/bin/git-getlock` and made executable (`chmod 755 /usr/local/bin/git-getlock`):
```
#!/bin/sh

git lfs lock "$1" && git pull -p
```

Then I can run `git getlock file.txt` and it does the two steps one after another.  (This works because on my machine `/usr/local/bin` is in my `PATH` environment variable, so Git looks for a `git-getlock` executable there when I run `git getlock`.)

You could also just create a script alias like `git-getlock` for yourself in Bash or whatever shell you're using, if you're on Unix.  If you're on Windows, PowerShell has lots of scripting facilities as well, for instance.

**Comment by SalmaFarghaly** [comment]
@chrisd8088 Thank you for your help.
Aside question, Is there a way to know which files are included in the push and have been locked by me in current checked out branch when running "git push" command ?

**Comment by chrisd8088** [comment, marked as answer]
Again, I think you'd have to write a script to help you out there.  You could use the output of `git lfs locks` (maybe with the `--json` output), parse that, and compare it against the output of something like `git diff --stat --cached <remote>/<branch>`, looking for common entries.
